### ISSUE TITLE
[Sampling] Optimize Top-P (Nucleus) Sampling Implementation for Production Performance

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#380]
## Problem Description

The current top-p (nucleus) sampling implementation in BitNet.rs, while functional, has several opportunities for optimization that are critical for production inference workloads. The implementation exists in two locations with different approaches, suggesting potential inconsistencies and performance bottlenecks.

## Environment

- **Affected Files**: 
  - `crates/bitnet-inference/src/sampling.rs` (lines 177-216)
  - `crates/bitnet-inference/src/generation/sampling.rs` (lines 173-204)
- **Component**: Sampling Strategy / Text Generation
- **Impact**: Performance-critical path for autoregressive generation
- **Build Configuration**: All features (`cpu`, `gpu`)

### Issues Identified

1. **Dual Implementation Inconsistency**: Two different top-p implementations exist:
   - `sampling.rs`: Direct probability array manipulation 
   - `generation/sampling.rs`: Candle tensor-based approach with softmax conversion

2. **Performance Bottlenecks**:
   - Repeated sorting operations for every sampling step
   - Multiple vector allocations and transformations
   - Inefficient cumulative probability calculation
   - Tensor conversion overhead in generation module

3. **Memory Inefficiency**:
   - Creates full-size filtered arrays even when nucleus is small
   - No reuse of sorted indices across sampling steps
   - Potential for large memory allocation spikes

### Code Analysis

**Current implementation in `sampling.rs`:**
```rust
fn apply_top_p(&self, probabilities: &[f32], p: f32) -> Vec<f32> {
    // Creates indices sorted by probability (descending)
    let mut indices: Vec<usize> = (0..probabilities.len()).collect();
    indices.sort_by(|&a, &b| probabilities[b].partial_cmp(&probabilities[a]).unwrap());
    
    // Find cutoff point - O(n) search
    let mut cumulative_prob = 0.0;
    // ... rest of implementation
}
```

**Performance Issues:**
- Sorting entire vocabulary every time: O(n log n)
- Full array allocation regardless of nucleus size
- No caching of frequently accessed probability distributions

## Root Cause Analysis

1. **Algorithmic Complexity**: Current O(n log n) sorting for every token generation
2. **Memory Patterns**: Inefficient allocation patterns causing cache misses
3. **Tensor Overhead**: Unnecessary conversions between tensor and array formats
4. **Code Duplication**: Maintaining two implementations increases maintenance burden

### Primary Approach: Optimized Nucleus Sampling

```rust
/// Optimized top-p sampling with performance enhancements
struct OptimizedNucleusSampler {
    // Pre-allocated working memory
    indices_buffer: Vec<usize>,
    cumulative_buffer: Vec<f32>,
    // Cached sorting for stable vocabularies
    last_sorted_indices: Option<Vec<usize>>,
    vocab_size: usize,
}

impl OptimizedNucleusSampler {
    /// Fast top-p sampling with memory reuse and early termination
    fn sample_nucleus(&mut self, logits: &[f32], p: f32) -> Result<Vec<f32>> {
        // 1. Use partial sorting (nth_element) instead of full sort
        // 2. Early termination when cumulative >= p
        // 3. Reuse buffers to minimize allocations
        // 4. Vectorized operations where possible
    }
    
    /// Adaptive threshold selection for common p values
    fn adaptive_threshold(&mut self, logits: &[f32], p: f32) -> f32 {
        // Fast path for common p values (0.9, 0.95, etc.)
        // Use binary search or approximate methods for threshold finding
    }
}
```

### Alternative Approaches

1. **Hardware-Accelerated Sampling**: 
   - GPU-based nucleus sampling for large vocabularies
   - SIMD optimization for CPU-based sorting

2. **Approximate Nucleus Sampling**:
   - Statistical approximation for very large vocabularies
   - Dynamic precision based on generation context

3. **Unified Implementation**:
   - Single optimized implementation serving both modules
   - Configurable backend (CPU/GPU) selection

### Phase 1: Consolidation (Priority: High)
- [ ] **Unify sampling implementations** into single optimized module
- [ ] **Benchmark current performance** with realistic workloads
- [ ] **Create performance regression tests** for sampling operations
- [ ] **Document sampling strategy selection** guidelines

### Phase 2: Optimization (Priority: High)  
- [ ] **Implement partial sorting** using `select_nth_unstable`
- [ ] **Add memory buffer reuse** to minimize allocations
- [ ] **Optimize cumulative probability calculation** with SIMD
- [ ] **Add adaptive thresholding** for common p values

### Phase 3: Advanced Features (Priority: Medium)
- [ ] **GPU-accelerated nucleus sampling** for large vocabularies
- [ ] **Approximate sampling modes** for extreme performance requirements
- [ ] **Dynamic vocabulary pruning** based on context
- [ ] **Sampling strategy auto-selection** based on performance metrics

### Phase 4: Integration (Priority: Medium)
- [ ] **Update autoregressive generator** to use optimized sampler
- [ ] **Add performance monitoring** for sampling operations
- [ ] **Create sampling strategy benchmarks** for different model sizes
- [ ] **Document performance characteristics** and tuning guidelines

### Performance Tests
```rust
#[bench]
fn bench_nucleus_sampling_large_vocab() {
    // Test with realistic 50k+ vocabulary sizes
    // Measure latency, memory usage, and throughput
}

#[bench] 
fn bench_nucleus_sampling_different_p_values() {
    // Test common p values: 0.1, 0.5, 0.9, 0.95, 0.99
    // Verify optimization effectiveness across range
}
```

### Correctness Tests
```rust
#[test]
fn test_nucleus_sampling_correctness() {
    // Cross-validate against reference implementation
    // Verify probability distribution properties
    // Test edge cases (p=0.0, p=1.0, empty logits)
}
```

### Integration Tests
```rust
#[test]
fn test_autoregressive_generation_with_optimized_sampling() {
    // End-to-end generation quality verification
    // Performance regression detection
}
```

### Performance Benefits
- **Latency Reduction**: 40-60% improvement in sampling time for large vocabularies
- **Memory Efficiency**: 30-50% reduction in peak memory usage during generation
- **Throughput**: 20-30% improvement in tokens/second for sampling-heavy workloads

### Quality Considerations
- **Deterministic Results**: Maintain exact reproducibility with seed control
- **Numerical Stability**: Preserve precision for edge cases and extreme p values
- **Compatibility**: Maintain API compatibility with existing generation code

## Acceptance Criteria

- [ ] **Single unified implementation** replaces dual sampling modules
- [ ] **Performance improvement**: >30% faster sampling for vocab_size > 10k
- [ ] **Memory reduction**: <50% peak memory usage during nucleus sampling
- [ ] **Correctness maintained**: All existing tests pass with identical outputs
- [ ] **Documentation updated**: Performance characteristics and usage guidelines
- [ ] **Benchmarks included**: Comprehensive performance test suite
- [ ] **GPU compatibility**: Optimized implementation works on CUDA/Metal backends

## Related Issues

This optimization will impact:
- Autoregressive generation performance
- Memory usage optimization  
- Production inference server scalability
- GPU kernel optimization

## Definition of Done

1. ✅ **Implementation complete** with optimized nucleus sampling
2. ✅ **Performance benchmarks** demonstrate >30% improvement
3. ✅ **All tests passing** including cross-validation with reference
4. ✅ **Documentation updated** with performance guidelines
5. ✅ **Code review completed** with focus on numerical stability
6. ✅ **Integration testing** verified in production inference scenarios

---

**Priority**: High - Core performance optimization for production inference
**Complexity**: Medium - Requires careful optimization without breaking correctness
**Timeline**: 2-3 weeks for complete implementation and testing
